plugins/experimental/wasm/wasm_main.cc

```diff
@@ -46,6 +46,250 @@ struct WasmInstanceConfig {
+      }
+    }
+
+    TSDebug(WASM_DEBUG_TAG, "[%s] retrieving returns from wasm handler function and pass back to ATS", __FUNCTION__);
+    if ((status == proxy_wasm::FilterDataStatus::Continue) ||
+        ((status == proxy_wasm::FilterDataStatus::StopIterationAndBuffer ||
+          status == proxy_wasm::FilterDataStatus::StopIterationAndWatermark) &&
+         eos && !towrite)) {
+      res = c->getTransformResult(&res_len);
+
+      if (res && res_len > 0) {
+        if (!ti->output_vio) {
+          if (eos && !towrite) {
+            ti->output_vio = TSVConnWrite(output_conn, contp, ti->output_reader, res_len); // HttpSM go on
+          } else {
+            ti->output_vio = TSVConnWrite(output_conn, contp, ti->output_reader, ti->downstream_bytes); // HttpSM go on
+          }
+        }
+
+        TSIOBufferWrite(ti->output_buffer, res, res_len);
+        ti->total  += res_len;
+        write_down  = 1;
+      }
+
+      c->clearTransformResult();
+    }
+
+    if (status == proxy_wasm::FilterDataStatus::StopIterationNoBuffer) {
+      c->clearTransformResult();
+    }
+
+    if (eos && !towrite) { // EOS
+      eos = 1;
+      break;
+    }
+
+  } while (towrite > 0);
+
+  if (eos && !ti->output_vio) {
+    ti->output_vio = TSVConnWrite(output_conn, contp, ti->output_reader, 0);
+  }
+
+  if (write_down || eos) {
+    TSVIOReenable(ti->output_vio);
+  }
+
+  if (toread > input_avail) { // upstream not finished.
+    if (eos) {
+      TSVIONBytesSet(ti->output_vio, ti->total);
+      if (empty_input == 0) {
+        TSContCall(TSVIOContGet(input_vio), TS_EVENT_VCONN_EOS, input_vio);
+      }
+    } else {
+      if (empty_input == 0) {
+        TSContCall(TSVIOContGet(input_vio), TS_EVENT_VCONN_WRITE_READY, input_vio);
+      }
+    }
+  } else { // upstream is finished.
+    TSVIONBytesSet(ti->output_vio, ti->total);
+    if (empty_input == 0) {
+      TSContCall(TSVIOContGet(input_vio), TS_EVENT_VCONN_WRITE_COMPLETE, input_vio);
+    }
+  }
+
+  return 0;
+}
+
+static int
+transform_entry(TSCont contp, TSEvent ev, void *edata)
+{
+  int event;
+  TSVIO input_vio;
+  ats_wasm::TransformInfo *ti;
+
+  event = (int)ev;
+  ti    = (ats_wasm::TransformInfo *)TSContDataGet(contp);
+
+  TSDebug(WASM_DEBUG_TAG, "[%s] begin transform entry", __FUNCTION__);
+  if (TSVConnClosedGet(contp)) {
+    delete ti;
+    TSContDestroy(contp);
+    return 0;
+  }
+
+  TSDebug(WASM_DEBUG_TAG, "[%s] checking event inside transform entry", __FUNCTION__);
+  switch (event) {
+  case TS_EVENT_ERROR:
+    TSDebug(WASM_DEBUG_TAG, "[%s] event error", __FUNCTION__);
+    input_vio = TSVConnWriteVIOGet(contp);
+    TSContCall(TSVIOContGet(input_vio), TS_EVENT_ERROR, input_vio);
+    break;
+
+  case TS_EVENT_VCONN_WRITE_COMPLETE:
+    TSDebug(WASM_DEBUG_TAG, "[%s] event vconn write complete", __FUNCTION__);
+    TSVConnShutdown(TSTransformOutputVConnGet(contp), 0, 1);
+    break;
+
+  case TS_EVENT_VCONN_WRITE_READY:
+  default:
```

Comment:
does `TS_EVENT_VCONN_EOS` not have a different course of action or is it assumed that it will eventually go to `TS_EVENT_VCONN_WRITE_COMPLETE`

Reply:
I think it will be TS_EVENT_VCONN_WRITE_COMPLETE eventually. 
I am following the null transform example plugin on the event handling

Reply:
actually from `transform_handler` logic (in my previous comment), I don't think `TS_EVENT_VCONN_EOS` would even be called